### ISSUE TITLE
Add support for native, explicit Python 3.5 coroutines

### ISSUE THREAD

**Issue by thedrow** [gevent/gevent#641]
See [PEP-0492](https://www.python.org/dev/peps/pep-0492/) for details.
This will allow to use geven in both explicit and implicit mode, and prefering the explicit style for new code which is more Pythonic.
It will also keep us in par with other, future implementations that will use asyncio or asyncio with gevent.

**Comment by jamadden** [comment]
@fantix has a Python 3 asyncio event loop at https://github.com/decentfox/tulipcore. I'm not sure it's reasonable or possible to do any more than that.

**Comment by thedrow** [comment]
The new coroutines are not asyncio specific as far as I can tell. They do integrate with asyncio but can be used without it.

**Comment by squeaky-pl** [comment]
I think it's interesting idea but maybe instead of being part of gevent could it be a separate project that depends on gevent?

**Comment by fantix** [comment]
There's also [aiogevent](https://bitbucket.org/haypo/aiogevent).

**Comment by thedrow** [comment]
@squeaky-pl The whole point is that we do not need greenlet anymore if we're using the explicit async/await keywords.
That means that greenlet will become an optional dependency in the feature as more and more APIs will implement the explicit async style.
Explicit behaviour is out of scope of gevent right now due to most APIs not supporting a way to run async without monkeypatching. That might change in the future with the new async/await keywords.
I think that the project must keep up with the changes in Python or it will die and be replaced with another component that does.
More over, if we do support 3.5 coroutines, supporting asyncio is a breeze. You don't need any adapter for the async operations since they are native Python constructs. All you need is an events loop.

**Comment by Ivoz** [comment]
@thedrow "implementing gevent core upon python 3.5 coroutines" (which is what making `greenlet` optional would involve) would be a complete rewrite of the entire guts of the project. This would make for a gevent 2.0 if anything. Incase you hadn't figured out the scale of the feature you're asking.

**Comment by thedrow** [comment]
I did not say it would be easy. The question is if there are valid points to not implement it inside the gevent core.

**Comment by thedrow** [comment]
Note that https://github.com/dabeaz/curio exists and implements it's own task queue.

**Comment by doublereedkurt** [comment]
Not to resurrect a very old thread, but since this hasn't been closed as a heavy user of gevent for the last four years: implicit coroutine switching is really core for code I write on gevent.

If a future version of gevent switched to requiring explicit coroutines I would not upgrade.  (Aside from how hard it is to implement from a user perspective it means throwing away lots of code.)

**Comment by jamadden** [comment]
I don't think we have any plans to make gevent require explicit coroutine switches, so you can rest easy :)

On thing that might be possible and would lead to closer integration would be to share the underlying event loop. I know that libuv can be plugged in as the asyncio event loop, and it might be possible to share the same loop between gevent and asyncio in that fashion.

**Comment by doublereedkurt** [comment]
That makes a lot of sense.  I've helped co-workers use the -- unfortunately abandoned -- gevent reactor (https://github.com/ivee/geventreactor) to run Scrapy inside a gevent web service without having to use separate threads.  gevent event.Event's and Twisted Deferred's can mix freely.  I was expecting it to be painful but actually we were up and running in an afternoon.

Explicit callbacks is lowest common denominator -- it is very easy to build a Twisted reactor on top of gevent, impossible to build a gevent hub / loop out of Twisted.

**Comment by jamadden** [comment]
Since this discussion has stalled out, and it's unlikely we'll ever require explicit coroutine switching, I'm closing for triage of 1.2a1. Once the libuv work lands (or even before) we can reopen discussion about integrating event loops with asyncio (though I read recently on one of the python mailing lists in the discussion about closing iterators that implementing the asyncio scheduler is considered to be black magic and they weren't expecting many such implementations, so that may be harder than anticipated).
